// repatch/lib/store.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./thunk"));
var Store = (function () {
    function Store(initialState) {
        var _this = this;
        this.listeners = [];
        this.getState = function () { return _this.state; };
        this.dispatch = function (reducer) {
            if (typeof reducer !== 'function')
                throw new Error('Reducer is not a function: dispatch takes only reducers as functions.');
            _this.state = reducer(_this.state);
            _this.listeners.forEach(function (listener) { return listener(); });
            return _this.state;
        };
        this.subscribe = function (listener) {
            if (typeof listener !== 'function')
                throw new Error('Listener is not a function: subscribe takes only listeners as functions.');
            _this.listeners = _this.listeners.concat([listener]);
            return function () { return (_this.listeners = _this.listeners.filter(function (lis) { return lis !== listener; })); };
        };
        this.addMiddleware = function () {
            var middlewares = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                middlewares[_i] = arguments[_i];
            }
            if (middlewares.some(function (middleware) { return typeof middleware !== 'function'; }))
                throw new Error('Middleware is not a function: addMiddleware takes only middlewares as functions.');
            middlewares.forEach(function (middleware) { return (_this.dispatch = middleware(_this)(_this.dispatch)); });
            return _this;
        };
        this.state = initialState;
    }
    return Store;
}());
exports.default = Store;
